src/reachability/ai2.jl

```diff
@@ -18,25 +32,49 @@ Reachability analysis using split and join.
 Sound but not complete.
 
 # Reference
-T. Gehr, M. Mirman, D. Drashsler-Cohen, P. Tsankov, S. Chaudhuri, and M. Vechev,
+[1] T. Gehr, M. Mirman, D. Drashsler-Cohen, P. Tsankov, S. Chaudhuri, and M. Vechev,
 "Ai2: Safety and Robustness Certification of Neural Networks with Abstract Interpretation,"
 in *2018 IEEE Symposium on Security and Privacy (SP)*, 2018.
+
+## Note
+Efficient over-approximation of intersections and unions involving zonotopes relies on Theorem 3.1 of
+
+[2] Singh, G., Gehr, T., Mirman, M., Püschel, M., & Vechev, M. (2018). Fast
+and effective robustness certification. In Advances in Neural Information
+Processing Systems (pp. 10802-10813).
 """
-struct Ai2 <: Solver end
+struct Ai2{T<:Union{Hyperrectangle, Zonotope, HPolytope}} <: Solver end
+
+Ai2() = Ai2{Zonotope}()
+const Ai2h = Ai2{HPolytope}
+const Ai2z = Ai2{Zonotope}
+const Box = Ai2{Hyperrectangle}
 
 function solve(solver::Ai2, problem::Problem)
     reach = forward_network(solver, problem.network, problem.input)
     return check_inclusion(reach, problem.output)
 end
 
-forward_layer(solver::Ai2, layer::Layer, inputs::Vector{<:AbstractPolytope}) = forward_layer.(solver, layer, inputs)
+forward_layer(solver::Ai2, L::Layer, inputs::Vector) = forward_layer.(solver, L, inputs)
 
-function forward_layer(solver::Ai2, layer::Layer, input::AbstractPolytope)
-    outlinear = affine_map(layer, input)
-    relued_subsets = forward_partition(layer.activation, outlinear) # defined in ExactReach
+function forward_layer(solver::Ai2h, L::Layer, input::AbstractPolytope)
+    Ẑ = affine_map(L, input)
+    relued_subsets = forward_partition(L.activation, Ẑ) # defined in reachability.jl
     return convex_hull(relued_subsets)
 end
 
+# method for Zonotope and Hyperrectangle, if the input set isn't a Zonotope
+function forward_layer(solver::Ai2, L::Layer, input::AbstractPolytope)
+    X = overapproximate(input, Hyperrectangle)
+    return forward_layer(solver, L, X)
+end
+
+function forward_layer(solver::Ai2{T}, L::Layer, input::AbstractZonotope) where T<:Union{Hyperrectangle, Zonotope}
+    Ẑ = affine_map(L, input)
+    return overapproximate(Rectification(Ẑ), T)
+end
+
+
 # extend lazysets convex_hull to a vector of polytopes
 function LazySets.convex_hull(sets::Vector{<:AbstractPolytope}; backend = CDDLib.Library())
```

Comment:
In LazySets there is a function that does this: https://github.com/JuliaReach/LazySets.jl/pull/2215

Reply:
Not quite. It returns a vector of vertices and we want a Polytope (we could do VPolytope(...) at the end though). Also, in hrep, I think it's more efficient not to go to vrep, no?

Reply:
The problem here is that you are doing a `convex_hull` every loop, while that implementation only does once, so, if there are more than two polytopes, I think that it will be faster even after the conversion, and it will scale much better with more polytopes and dimensions.

Reply:
Hmm, this could be. I didn't think so since I assumed `convex_hull` of two polytopes of the same type does not convert, and therefore it should be more efficient. Since the code in LazySets defers to Polyhedra, do you know if this is the case or not?

Reply:
The function used is `removehredundancy()` [here](https://github.com/JuliaReach/LazySets.jl/blob/77e9f6c30f8f49fbde6fb8137a0ddfcd820eda9b/src/Sets/HPolyhedron.jl#L451), so it does not convert to v-rep.
I did a little benchmark with two polygons:
```julia
julia> @btime convert(HPolytope, VPolytope(convex_hull(UnionSetArray([$H, $H]))));
  331.000 ns (4620 allocations: 310.05 KiB)
julia> @btime convex_hull($H, $H);
  1.072 ms (14202 allocations: 864.63 KiB)
```
So it seems to me that the implementation in LazySets is faster even in 2D and two polytopes

Reply:
Computing `Pch` in Polyhedra requires vrep I think. The Polyhedra documentation suggests it is only possible in vrep
https://juliapolyhedra.github.io/Polyhedra.jl/latest/utilities/#Polyhedra.convexhull

Therefore your method is better after all I guess

Reply:
By the way, I have recently noticed printouts from the LP solver used by LazySets (or lower in the stack) which were not there in prior package versions. Do you know which issue/s these new messages are related to? An example is:
```julia
julia> convex_hull(UnionSetArray([hp, hp]))
glp_simplex: unable to recover undefined or non-optimal solution
glp_simplex: unable to recover undefined or non-optimal solution
glp_simplex: unable to recover undefined or non-optimal solution
glp_simplex: unable to recover undefined or non-optimal solution
glp_simplex: unable to recover undefined or non-optimal solution
glp_simplex: unable to recover undefined or non-optimal solution
3-element Array{Array{Float64,1},1}:
 [-0.36626701868929534, 1.1382211150524333, -0.6787710912863433]
 [0.36968831822766857, -0.7100424633428055, -0.1920207505884401]
 [-1.9245657296871925, 0.42914282672330284, -0.7531344058289536]
```

Reply:
I don't know where they originate but there are issues that mentions it(https://github.com/JuliaReach/LazySets.jl/issues/2278#issuecomment-664180892 and https://github.com/JuliaReach/LazySets.jl/issues/861#issuecomment-660503226)